### DISCUSSION TITLE
Git commands as if condition of a step in workflow

### DISCUSSION THREAD

**Discussion by navankurverma** [community/community#27148] [Actions]
<p>Hello, I want to only run the ‘Commit and Push’ step if there are some changes in the repo.<br>
What I tried to do:</p>
<pre><code class="lang-auto">- name: Commit and Push
      if: 'git diff --exit-code' 
   run: | 
       git add -u
       git commit -m "github-actions added date data"
       git push --force https://${GITHUB_ACTOR}:${{secrets.GITHUB_TOKEN}}@github.com/${GITHUB_REPOSITORY}.git HEAD:${REMOTE_BRANCH} 
</code></pre>
<p>But it gives error (I tried to search online whether an git command can be used as condition or not, I know it works in linux terminal but as error suggests github action does not seem to support it):</p>
<pre><code class="lang-auto">Invalid workflow file 
The workflow is not valid. .github/workflows/main.yml (Line: 45, Col: 11): Unrecognized named-value: 'git'. Located at position 1 within expression: git diff --exit-code 
</code></pre>
<p>If this is not the correct way, please recommend any other option to achieve the same?</p>
<p>Thanks</p>

**Comment by airtower-luna** [comment]
<p><code>if:</code> is no shell, so no, running commands there is not supported. What you can do is <a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="nofollow noopener">set an environment variable or output variable</a> in a previous step, and check that. Example with environment variable:</p>
<pre><code class="lang-yaml">- name: check for changes
  run: |
    if git diff --exit-code; then
      echo "::set-env name=changes_exist::true"
    else
      echo "::set-env name=changes_exist::false"
    fi
- name: Commit and Push
  if: ${{ env.changes_exist == 'true' }}
  run: | 
    git add -u
    git commit -m "github-actions added date data"
    git push --force https://${GITHUB_ACTOR}:${{secrets.GITHUB_TOKEN}}@github.com/${GITHUB_REPOSITORY}.git HEAD:${REMOTE_BRANCH}
</code></pre>
<p>Also keep in mind that indentation is part of YAML syntax, uneven indentation like in your example might cause additional errors.</p>

**Comment by navankurverma** [comment]
<p>Hi Luna,</p>
<p>Thanks for the approach. I tried but still getting error:</p>
<pre><code class="lang-auto">    - name: check for changes
      run: |
        if git diff --exit-code; then
            echo "::set-env name=changes_exist::true"
        else
            echo "::set-env name=changes_exist::false"   
        fi
        
    - name: Commit and Push Only if file updated
      if: env.changes_exist == 'true'
      run: |
        git add -u
        git commit -m "github-actions added date data"
        git push --force https://${GITHUB_ACTOR}:${{secrets.GITHUB_TOKEN}}@github.com/${GITHUB_REPOSITORY}.git HEAD:${REMOTE_BRANCH}
</code></pre>
<p>I tried many permutations and in one of those the step was actually skipped but job failed.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/c23463d81c5b8574bcb95b0e78b14a171816a209.png" data-download-href="/uploads/short-url/rI0Sd2l3UQq1rRUgUZW67wvqljX.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/11485752/181102062-7adda901-50a8-4302-82e0-ee168f25ee38.png" alt="image" data-base62-sha1="rI0Sd2l3UQq1rRUgUZW67wvqljX" width="441" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/c/c23463d81c5b8574bcb95b0e78b14a171816a209_2_441x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/c/c23463d81c5b8574bcb95b0e78b14a171816a209_2_661x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/c/c23463d81c5b8574bcb95b0e78b14a171816a209.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/c23463d81c5b8574bcb95b0e78b14a171816a209_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">814×921 41 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>What am I missing here?</p>
<p>Thanks</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Sorry, I put <code>true</code> and <code>false</code> the wrong way around in my example. According to the documentation <code>git diff --exit-code</code> will return 0 if there are <em>no</em> changes, and in Bash 0 means true, contrary to C… :sweat_smile:</p>
<p>So the check should be:</p>
<pre><code class="lang-yaml">- name: check for changes
  run: |
    if git diff --exit-code; then
      echo "::set-env name=changes_exist::false"
    else
      echo "::set-env name=changes_exist::true"
    fi
</code></pre>
<p>I hope this works, at least the mixup explains why you got “nothing to commit” in the first screenshot.</p>

**Comment by navankurverma** [comment]
<p>No, dont be sorry. Programming with such things is confusing.</p>
<p>Thanks a lot, it worked.</p>
<p><img src="https://user-images.githubusercontent.com/11485752/181102066-cd4c78c3-8567-40f3-b494-80e82ef5d72f.png" alt="image" data-base62-sha1="jNhBMAqGZNmWJq7fL9FdiOtl9uP" width="452" height="284"></p>
